components/ai_chat/resources/page/components/main/index.tsx

```diff
@@ -85,6 +89,21 @@ function Main() {
     }
   }
 
+  const handleScroll = (e: React.UIEvent<HTMLDivElement>) => {
+    const el = e.currentTarget
+    scrollPos.current.isAtBottom = Math.abs(el.scrollHeight - el.clientHeight - el.scrollTop) < SCROLL_BOTTOM_THRESHOLD
+  }
+
+  const handleLastElementHeightChange = () => {
+    if (!scrollerElement) {
+      return
+    }
+
+    if (scrollPos.current.isAtBottom) {
```

Comment:
Why can't we just here check if `scrollerElement` is at the bottom and not need to have `handleScroll` and `scrollPos`?

Reply:
i did try that initially, but it looks like there's a timing issue. basically, computing `isAtBottom` in the same call stack is flawed. additionally, it doesn't monitor when the user has scrolled to the bottom, which we need for auto scrolling behavior.

Reply:
Well, perhaps we can overcome the potential performance issue another way then. I'm seeing handleScroll called on every scroll when it's not neccessary until the content of the "last element" changes.
I'm assuming your issue when you tried this initially is because `handleLastElementHeightChange` is called immediately when the content of the react properties change but before they are rendered (and definitely before the html layout is done). So couldn't that be fixed with a `requestAnimationFrame`?

Reply:
i have also tried using `rAF` to sync with browser's rendering within the same callstack. however, we are still off.

still, this approach doesn't resolve the need to continuously monitor the user's scroll while the generation is occurring. if the user decides to scroll back to the bottom, we need to switch back to the scroll-to-bottom behavior.

i agree that `handleScroll` doesn't really need to be called on every tick. we can overcome this issue by slightly throttling.

Reply:
Yes, throttle at the very least. Or `useLayoutEffect`?

Reply:
actually, we do need to monitor every tick. throttling can be unreliable because if the user is at the bottom, it may compute positions at a later time, potentially missing the opportunity to autoscroll.

we could run the scroll handler in `rAF` and make it tick with the monitor's refresh rate. ex. https://codepen.io/dcorb/pen/pgOKKw - potentially lowers the number of ticks compared to before but not significantly as throttling.

Reply:
another idea: what if we run or attach the scroll handler only if `context.isGenerating` ?

Reply:
changed to only monitor when it's generating